geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/KryoLazyDensityUtils.scala

```diff
@@ -18,48 +18,46 @@ import org.locationtech.geomesa.utils.geotools.GridSnap
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 import org.opengis.filter.expression.Expression
+import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
 
 import scala.collection.mutable
 
-trait KryoLazyDensityUtils {
+class KryoLazyDensityUtils(options: Map[String, String], sft: SimpleFeatureType) {
 
   import KryoLazyDensityUtils._
-  var geomIndex: Int = -1
+
+  val geomIndex: Int = Option(sft.getGeomIndex).getOrElse(-1)
   // we snap each point into a pixel and aggregate based on that
-  var gridSnap: GridSnap = null
-  var writeGeom: (SimpleFeature, DensityResult) => Unit = null
-
-  def initialize(options: Map[String, String], sft: SimpleFeatureType): DensityResult = {
-    import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
-    geomIndex = sft.getGeomIndex
-    gridSnap = {
-      val bounds = options(ENVELOPE_OPT).split(",").map(_.toDouble)
-      val envelope = new Envelope(bounds(0), bounds(1), bounds(2), bounds(3))
-      val Array(width, height) = options(GRID_OPT).split(",").map(_.toInt)
-      new GridSnap(envelope, width, height)
+  val gridSnap = {
+    val bounds = options(ENVELOPE_OPT).split(",").map(_.toDouble)
+    val envelope = new Envelope(bounds(0), bounds(1), bounds(2), bounds(3))
+    val Array(width, height) = options(GRID_OPT).split(",").map(_.toInt)
+    new GridSnap(envelope, width, height)
+  }
+  // function to get the weight from the feature - defaults to 1.0 unless an attribute/exp is specified
+  val weightIndex = options.get(WEIGHT_OPT).map(sft.indexOf).getOrElse(-2)
+
+  def newDensityResult: DensityResult = mutable.Map.empty[(Int, Int), Double]
+
+  def weightFn(sf: SimpleFeature): Double = {
+    if (weightIndex == -2) {
+      1.0
+    } else if (weightIndex == -1) {
+      val expression = ECQL.toExpression(options(WEIGHT_OPT))
+      getWeightFromExpression(expression)(sf)
+    } else if (sft.getDescriptor(weightIndex).getType.getBinding == classOf[java.lang.Double]) {
+      getWeightFromDouble(weightIndex)(sf)
+    } else {
+      getWeightFromNonDouble(weightIndex)(sf)
     }
+  }
 
-    // function to get the weight from the feature - defaults to 1.0 unless an attribute/exp is specified
-    val weightIndex = options.get(WEIGHT_OPT).map(sft.indexOf).getOrElse(-2)
-    val weightFn: (SimpleFeature) => Double =
-      if (weightIndex == -2) {
-        (_) => 1.0
-      } else if (weightIndex == -1) {
-        val expression = ECQL.toExpression(options(WEIGHT_OPT))
-        getWeightFromExpression(expression)
-      } else if (sft.getDescriptor(weightIndex).getType.getBinding == classOf[java.lang.Double]) {
-        getWeightFromDouble(weightIndex)
-      } else {
-        getWeightFromNonDouble(weightIndex)
-      }
-
-    writeGeom = if (sft.isPoints) {
-      (sf, result) => writePoint(sf, weightFn(sf), result)
+  def writeGeom(sf: SimpleFeature, result: DensityResult): Unit = {
```

Comment:
in general these changes look fine - eventually it might be cleaner to create a companion object that does the setup and configuration, and then a simpler class implementation (or multiple impls, one for points, non-points, etc)